### PR TITLE
Reauthentication

### LINKED ISSUE #119: Support reauthentication (JS SDK)
- Add `getIDTokenHint`
- Add `canReauthenticate`
- Refresh ID token with custom grant type
- Add `reauthenticate`

### PR DESCRIPTION
**Pull request by louischan-oursky**
fix #119

### CODE REVIEW

File: packages/authgear-web/src/container.ts [hunk 1]
```diff
@@ -270,26 +272,55 @@ export class WebContainer {
...
-   * Finish authorization
+   * Start reauthentication by redirecting to the authorization endpoint.
+   */
+  async startReauthentication(options: ReauthenticateOptions): Promise<void> {
+    await this.baseContainer._refreshIDToken();
+
+    const idToken = this.getIDTokenHint();
+    if (idToken == null || !this.canReauthenticate()) {
+      throw new Error(
+        "You can only trigger reauthentication when canReauthenticate() returns true"
```
Comment (carmenlau) [inline packages/authgear-web/src/container.ts:291]: If we throw error here, do we expect developer should check `canReauthenticate` before calling reauthentication? If this is the case, seems we need some way to ensure `canReauthenticate` is up to date? e.g. User originally doesn't have mfa, and added mfa in the setting page later.

File: packages/authgear-core/src/base64.ts [hunk 2]
```diff
@@ -1,43 +1,18 @@
...
+    .replace(/=+$/, "");
 }
 
 /**
  * @internal
  */
-export function _encodeBase64URLFromByteArray(bytes: _ByteArray): string {
-  return base64URLEncArr(bytes);
+export function _base64URLDecode(s: string): ArrayBuffer {
+  return decode(s.replace(/-/g, "+").replace(/_/g, "/"));
```
Comment (carmenlau) [inline packages/authgear-core/src/base64.ts:17]: Is it necessary to add back tailing `=` like what we did in iOS sdk?

Comment (louischan-oursky) [reply packages/authgear-core/src/base64.ts:17]: I have checked the source code of the library we are using. [The padding is actually ignored.](https://github.com/niklasvh/base64-arraybuffer/blob/master/lib/base64-arraybuffer.js#L44) so I did not add the padding.

Comment (carmenlau) [reply packages/authgear-core/src/base64.ts:17]: ok, that's great. Could we put comments here so we won't confuse in the future? :)

**Pull request merged**